File: sdk/js/test/device.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { GraftDevice } from '../src/device.js'
import { MockTransport } from '../src/transports/mock.js'
import { wireEncode, wireDecode, makePacket } from '../src/core/wire.js'
import { CMD, EVT, PROTOCOL_VERSION, DS_TYPE } from '../src/core/constants.js'

/**
 * Build a HELLO_RESP packet that the mock device sends back.
 */
function makeHelloRespPayload(): Uint8Array {
  const parts: number[] = []
  // firmware_name: 16 bytes
  const name = new TextEncoder().encode('MockBoard')
  for (let i = 0; i < 16; i++) parts.push(i < name.length ? name[i] : 0)
  // version: 1.0.0
  parts.push(1, 0, 0)
  // mcu_id: 8 bytes
  for (let i = 0; i < 8; i++) parts.push(0)
  // ota_capable
  parts.push(0)
  // pin_count = 20
  parts.push(20)
  for (let i = 0; i < 20; i++) parts.push(0x0F) // DIN|DOUT|PWM|AIN
  // i2c, spi, uart
  parts.push(1, 1, 1)
  // max_payload: 512
  parts.push(0x00, 0x02)
  // module_count = 1 (servo)
  parts.push(1)
  parts.push(0) // module_id
  const modName = new TextEncoder().encode('servo')
  for (let i = 0; i < 8; i++) parts.push(i < modName.length ? modName[i] : 0)
  parts.push(1, 0) // version 1.0
  parts.push(0) // 0 pins claimed
  // datastream_count = 1
  parts.push(1)
  const dsName = new TextEncoder().encode('temperature')
  for (let i = 0; i < 16; i++) parts.push(i < dsName.length ? dsName[i] : 0)
  parts.push(DS_TYPE.FLOAT32)
  const unit = new TextEncoder().encode('celsius')
  for (let i = 0; i < 8; i++) parts.push(i < unit.length ? unit[i] : 0)
  parts.push(0) // not writable
  parts.push(0xFF) // no pin ref
  parts.push(0) // no retain
  return new Uint8Array(parts)
}

/**
 * Auto-respond mock: when the SDK sends a HELLO command,
 * respond with HELLO_RESP. When it sends any other command, respond ACK.
 */
function autoRespond(transport: MockTransport) {
  const origSend = transport.send.bind(transport)
  transport.send = async (packet: Uint8Array) => {
    await origSend(packet)
    // Decode the sent packet to get seq and type
    try {
      const decoded = wireDecode(packet)
      if (decoded.type === CMD.HELLO) {
        const resp = wireEncode(makePacket(EVT.HELLO_RESP, decoded.seq, makeHelloRespPayload()))
        transport.inject(resp)
      } else if (decoded.type === CMD.PING) {
        const resp = wireEncode(makePacket(EVT.PONG, decoded.seq))
        transport.inject(resp)
      } else {
        // ACK everything else
        const resp = wireEncode(makePacket(EVT.ACK, decoded.seq))
        transport.inject(resp)
      }
    } catch { /* ignore decode errors */ }
  }
}

describe('GraftDevice', () => {
  let transport: MockTransport

  beforeEach(() => {
    transport = new MockTransport()
    autoRespond(transport)
  })

  it('connects and parses HELLO_RESP', async () => {
    const device = await GraftDevice.connect(transport)
    expect(device.connected).toBe(true)
    expect(device.capabilities).not.toBeNull()
    expect(device.capabilities!.firmwareName).toBe('MockBoard')
    expect(device.capabilities!.firmwareVersion).toEqual([1, 0, 0])
    expect(device.capabilities!.pins.length).toBe(20)
    expect(device.capabilities!.modules.length).toBe(1)
    expect(device.capabilities!.modules[0].name).toBe('servo')
    expect(device.capabilities!.datastreams.length).toBe(1)
    expect(device.capabilities!.datastreams[0].name).toBe('temperature')
  })

  it('sends PING and receives PONG', async () => {
    const device = await GraftDevice.connect(transport)
    await device.ping() // should not throw
  })

  it('sets pin mode', async () => {
    const device = await GraftDevice.connect(transport)
    await device.pin(13).mode('output')

    // Verify a PIN_MODE command was sent
    const sent = transport.sentPackets
    const pinModePkt = sent.find(p => {
      try { return wireDecode(p).type === CMD.PIN_MODE } catch { return false }
    })
    expect(pinModePkt).toBeDefined()
    const decoded = wireDecode(pinModePkt!)
    expect(decoded.payload[0]).toBe(13) // pin
    expect(decoded.payload[1]).toBe(0x01) // OUTPUT
  })

  it('writes to pin', async () => {
    const device = await GraftDevice.connect(transport)
    await device.pin(13).write(1)

    const sent = transport.sentPackets
    const pinWritePkt = sent.find(p => {
      try { return wireDecode(p).type === CMD.PIN_WRITE } catch { return false }
    })
    expect(pinWritePkt).toBeDefined()
    const decoded = wireDecode(pinWritePkt!)
    expect(decoded.payload[0]).toBe(13)
    expect(decoded.payload[1]).toBe(1)
  })

  it('reads from pin', async () => {
    // Override auto-respond for PIN_READ to return a value
    transport.send = async (packet: Uint8Array) => {
      transport.sentPackets.push(new Uint8Array(packet))
      try {
        const decoded = wireDecode(packet)
        if (decoded.type === CMD.HELLO) {
          transport.inject(wireEncode(makePacket(EVT.HELLO_RESP, decoded.seq, makeHelloRespPayload())))
        } else if (decoded.type === CMD.PIN_READ) {
          // Return pin 0, value 512 (little-endian)
          const resp = wireEncode(makePacket(EVT.PIN_READ_RESP, decoded.seq, new Uint8Array([0, 0x00, 0x02])))
          transport.inject(resp)
        } else {
          transport.inject(wireEncode(makePacket(EVT.ACK, decoded.seq)))
        }
      } catch { /* ignore */ }
    }

    const device = await GraftDevice.connect(transport)
    const value = await device.pin(0).read()
    expect(value).toBe(512)
  })

  it('sends module command', async () => {
    const device = await GraftDevice.connect(transport)
    const servo = device.module('servo')
    await servo.cmd(0x02, new Uint8Array([90])) // write angle 90

    const sent = transport.sentPackets
    const modPkt = sent.find(p => {
      try { return wireDecode(p).type === CMD.MOD_CMD } catch { return false }
    })
    expect(modPkt).toBeDefined()
    const decoded = wireDecode(modPkt!)
    expect(decoded.payload[0]).toBe(0) // module_id
    expect(decoded.payload[1]).toBe(0x02) // cmd: write angle
    expect(decoded.payload[2]).toBe(90)
  })

  it('throws on unknown module', async () => {
    const device = await GraftDevice.connect(transport)
    expect(() => device.module('nonexistent')).toThrow('not found')
  })

  it('sends reset', async () => {
    const device = await GraftDevice.connect(transport)
    await device.reset()

    const sent = transport.sentPackets
    const resetPkt = sent.find(p => {
      try { return wireDecode(p).type === CMD.RESET } catch { return false }
    })
    expect(resetPkt).toBeDefined()
  })

  it('disconnects cleanly', async () => {
    const device = await GraftDevice.connect(transport)
    await device.disconnect()
    expect(device.connected).toBe(false)
    expect(device.capabilities).toBeNull()
  })

  it('handles NAK responses', async () => {
    transport.send = async (packet: Uint8Array) => {
      transport.sentPackets.push(new Uint8Array(packet))
      try {
        const decoded = wireDecode(packet)
        if (decoded.type === CMD.HELLO) {
          transport.inject(wireEncode(makePacket(EVT.HELLO_RESP, decoded.seq, makeHelloRespPayload())))
        } else {
          // NAK with INVALID_PIN error
          transport.inject(wireEncode(makePacket(EVT.NAK, decoded.seq, new Uint8Array([0x04]))))
        }
      } catch { /* ignore */ }
    }

    const device = await GraftDevice.connect(transport)
    await expect(device.pin(99).mode('output')).rejects.toThrow('NAK')
  })

  it('increments SEQ per command', async () => {
    const device = await GraftDevice.connect(transport)

    // HELLO was seq 0, so next commands start from 1
    await device.pin(13).mode('output')
    await device.pin(13).write(1)
    await device.pin(13).write(0)

    // Check SEQ values are incrementing
    const seqs = transport.sentPackets.map(p => {
      try { return wireDecode(p).seq } catch { return -1 }
    }).filter(s => s >= 0)

    // Should be sequential
    for (let i = 1; i < seqs.length; i++) {
      expect(seqs[i]).toBe((seqs[i - 1] + 1) & 0xFF)
    }
  })
})

File: sdk/go/device_test.go
package conduyt

import (
	"context"
	"encoding/binary"
	"testing"
	"time"

	"github.com/virgilvox/conduyt/sdk/go/transports"
)

// makeHelloRespPayload builds a binary HELLO_RESP payload for testing.
// Firmware: "MockBoard", version 1.0.0, 20 pins with 0x0F caps,
// 1 servo module, 1 temperature datastream.
func makeHelloRespPayload() []byte {
	var buf []byte

	// Firmware name (16 bytes, null-padded)
	name := make([]byte, 16)
	copy(name, "MockBoard")
	buf = append(buf, name...)

	// Version: 1.0.0
	buf = append(buf, 1, 0, 0)

	// MCU ID (8 bytes)
	buf = append(buf, 0xDE, 0xAD, 0xBE, 0xEF, 0x01, 0x02, 0x03, 0x04)

	// OTA capable
	buf = append(buf, 0x01)

	// 20 pins, each with caps = 0x0F
	buf = append(buf, 20)
	for i := 0; i < 20; i++ {
		buf = append(buf, 0x0F)
	}

	// I2C buses, SPI buses, UART count
	buf = append(buf, 1, 1, 1)

	// Max payload (uint16 LE)
	mp := make([]byte, 2)
	binary.LittleEndian.PutUint16(mp, 256)
	buf = append(buf, mp...)

	// 1 module: servo
	buf = append(buf, 1) // module count
	buf = append(buf, 0x01) // module ID
	modName := make([]byte, 8)
	copy(modName, "Servo")
	buf = append(buf, modName...)
	buf = append(buf, 1, 0) // version 1.0
	buf = append(buf, 1)    // 1 pin
	buf = append(buf, 9)    // pin 9

	// 1 datastream: temperature
	buf = append(buf, 1) // datastream count
	dsName := make([]byte, 16)
	copy(dsName, "temperature")
	buf = append(buf, dsName...)
	buf = append(buf, 0x02) // type: float
	dsUnit := make([]byte, 8)
	copy(dsUnit, "C")
	buf = append(buf, dsUnit...)
	buf = append(buf, 0x00) // not writable
	buf = append(buf, 0xFF) // pinRef = none
	buf = append(buf, 0x01) // retain = true

	return buf
}

// autoResponder starts a goroutine that watches for sent packets on the mock
// transport and injects appropriate responses. It must be started before
// calling device methods that send commands.
func autoResponder(mock *transports.MockTransport, responses map[byte]func(seq byte)) {
	go func() {
		seen := 0
		for {
			// Poll for new sent packets
			time.Sleep(1 * time.Millisecond)
			for seen < len(mock.SentPackets) {
				raw := mock.SentPackets[seen]
				seen++
				pkt, err := WireDecode(raw)
				if err != nil {
					continue
				}
				if fn, ok := responses[pkt.Type]; ok {
					fn(pkt.Seq)
				}
			}
		}
	}()
}

// injectResponse creates a wire-encoded response and injects it into the mock.
func injectResponse(mock *transports.MockTransport, evtType byte, seq byte, payload []byte) {
	pkt := MakePacket(evtType, seq, payload)
	mock.Inject(WireEncode(pkt))
}

func TestDeviceConnect(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	hello, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	if hello.FirmwareName != "MockBoard" {
		t.Errorf("expected firmware name MockBoard, got %s", hello.FirmwareName)
	}
	if hello.FirmwareVersion != [3]byte{1, 0, 0} {
		t.Errorf("expected version 1.0.0, got %v", hello.FirmwareVersion)
	}
	if len(hello.Pins) != 20 {
		t.Errorf("expected 20 pins, got %d", len(hello.Pins))
	}
	if len(hello.Modules) != 1 {
		t.Errorf("expected 1 module, got %d", len(hello.Modules))
	}
	if hello.Modules[0].Name != "Servo" {
		t.Errorf("expected module name Servo, got %s", hello.Modules[0].Name)
	}
	if len(hello.Datastreams) != 1 {
		t.Errorf("expected 1 datastream, got %d", len(hello.Datastreams))
	}
	if hello.Datastreams[0].Name != "temperature" {
		t.Errorf("expected datastream name temperature, got %s", hello.Datastreams[0].Name)
	}

	_ = dev.Close()
}

func TestDevicePing(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
		CmdPing: func(seq byte) {
			injectResponse(mock, EvtPong, seq, nil)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	err = dev.Ping(ctx)
	if err != nil {
		t.Fatalf("Ping failed: %v", err)
	}

	_ = dev.Close()
}

func TestPinMode(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
		CmdPinMode: func(seq byte) {
			injectResponse(mock, EvtACK, seq, nil)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	err = dev.Pin(13).Mode(ctx, PinModeOutput)
	if err != nil {
		t.Fatalf("PinMode failed: %v", err)
	}

	// Verify the sent PIN_MODE packet
	found := false
	for _, raw := range mock.SentPackets {
		pkt, err := WireDecode(raw)
		if err != nil {
			continue
		}
		if pkt.Type == CmdPinMode {
			if len(pkt.Payload) != 2 || pkt.Payload[0] != 13 || pkt.Payload[1] != PinModeOutput {
				t.Errorf("unexpected PIN_MODE payload: %v", pkt.Payload)
			}
			found = true
			break
		}
	}
	if !found {
		t.Error("PIN_MODE packet not found in sent packets")
	}

	_ = dev.Close()
}

func TestPinWrite(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
		CmdPinWrite: func(seq byte) {
			injectResponse(mock, EvtACK, seq, nil)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	err = dev.Pin(13).Write(ctx, 1)
	if err != nil {
		t.Fatalf("PinWrite failed: %v", err)
	}

	// Verify the sent PIN_WRITE packet
	found := false
	for _, raw := range mock.SentPackets {
		pkt, err := WireDecode(raw)
		if err != nil {
			continue
		}
		if pkt.Type == CmdPinWrite {
			if len(pkt.Payload) != 2 || pkt.Payload[0] != 13 || pkt.Payload[1] != 1 {
				t.Errorf("unexpected PIN_WRITE payload: %v", pkt.Payload)
			}
			found = true
			break
		}
	}
	if !found {
		t.Error("PIN_WRITE packet not found in sent packets")
	}

	_ = dev.Close()
}

func TestPinRead(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
		CmdPinRead: func(seq byte) {
			// PIN_READ_RESP: [pin, value_lo, value_hi]
			payload := make([]byte, 3)
			payload[0] = 13
			binary.LittleEndian.PutUint16(payload[1:3], 512)
			injectResponse(mock, EvtPinReadResp, seq, payload)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	val, err := dev.Pin(13).Read(ctx)
	if err != nil {
		t.Fatalf("PinRead failed: %v", err)
	}
	if val != 512 {
		t.Errorf("expected pin read value 512, got %d", val)
	}

	_ = dev.Close()
}

func TestDeviceClose(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	err = dev.Close()
	if err != nil {
		t.Fatalf("Close failed: %v", err)
	}

	if mock.Connected() {
		t.Error("transport should be disconnected after Close")
	}
}

func TestNAKError(t *testing.T) {
	mock := transports.NewMock()
	helloPayload := makeHelloRespPayload()

	responses := map[byte]func(seq byte){
		CmdHello: func(seq byte) {
			injectResponse(mock, EvtHelloResp, seq, helloPayload)
		},
		CmdPing: func(seq byte) {
			injectResponse(mock, EvtNAK, seq, []byte{ErrUnknownType})
		},
	}
	autoResponder(mock, responses)

	dev := NewDevice(mock, 2*time.Second)
	ctx := context.Background()

	_, err := dev.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}

	err = dev.Ping(ctx)
	if err == nil {
		t.Fatal("expected NAK error, got nil")
	}

	nakErr, ok := err.(*NAKError)
	if !ok {
		t.Fatalf("expected *NAKError, got %T: %v", err, err)
	}
	if nakErr.Code != ErrUnknownType {
		t.Errorf("expected NAK code 0x%02x, got 0x%02x", ErrUnknownType, nakErr.Code)
	}

	_ = dev.Close()
}
